scripts/pylib/twister/twisterlib/statuses.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+# Copyright (c) 2024 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
+Status classes to be used instead of str statuses.
+"""
+
+from enum import Enum
+
+
+# Status Assignment tree:
+#   parent <- child <=> parent is assigned to child in code.
+#
+#   HarnessStatus
+#   └ QEMUOutputStatus
+#     └ TestInstanceStatus
+#       ├ TestSuiteStatus
+#       └ TestCaseStatus
+
+
+class TestInstanceStatus(str, Enum):
+    def __str__(self):
+        return str(self.value)
+
+    # Values assigned directly
+    NONE = None  # to preserve old functionality
+    ERROR = 'error'
+    FAIL = 'failed'
+    FILTER = 'filtered'
+    PASS = 'passed'
+    SKIP = 'skipped'
+
+
+class TestSuiteStatus(str, Enum):
+    def __str__(self):
+        return str(self.value)
+
+    # Values assigned directly
+    NONE = None  # to preserve old functionality
+    FILTER = 'filtered'
+    PASS = 'passed'
+    SKIP = 'skipped'
+
+    # Values assigned via TestInstanceStatus
+    ERROR = 'error'
+    FAIL = 'failed'
+
+
+class TestCaseStatus(str, Enum):
+    def __str__(self):
+        return str(self.value)
+
+    # Values assigned directly
+    NONE = None  # to preserve old functionality
+    BLOCK = 'blocked'
+    ERROR = 'error'
+    FAIL = 'failed'
+    FILTER = 'filtered'
+    PASS = 'passed'
+    SKIP = 'skipped'
+    STARTED = 'started'
+
+
+class QEMUOutputStatus(str, Enum):
+    def __str__(self):
+        return str(self.value)
+
+    # Values assigned directly
+    NONE = None  # to preserve old functionality
+    BYTE = 'unexpected byte'
```

Comment:
This commit was planned only to isolate all the different statuses and their assignments, deduce what statuses are currently in use and separate them into a separate class. It was a descriptive change that describes how we currently approach statuses, just in an easier-to-read way.

Status-assignment code was borne out of a desire to protect the programmers from incorrect status assignments. Note that had we had that before, catching the strange QEMU statuses that honestly should just be an `ERROR` with a proper "reason" field could have been easier.

If we merge all the statuses into one, then why is there a `STARTED` status that is assignable for e.g. `Harness`? What does it mean for `Harness` to be `BLOCK`? Without separate status classes and without a separate property assignment code in status-having classes, we invite such errors. This could be resolved with a consistent status policy, where all statuses make sense on all classes that can have a status, but that would be even more changes to the status ecosystem in one big PR. At that point, it might be easier to revive #61414 directly.

Reply:
see https://github.com/zephyrproject-rtos/zephyr/pull/73358 for a proposal of how we should fix this

Reply:
I left the QEMUHandler here, if #73358 is merged first, it will be easier to incorporate.